pkg/proxy/provider/http/http.go

```
@@ -0,0 +1,164 @@
+							errChan <- err
+							return
+						}
+
+						configuration := p.buildConfiguration(ctx, data)
+						if configuration != nil {
+							configurationChan <- dynamic.Message{
+								ProviderName:  "http",
+								Configuration: configuration,
+							}
+						}
+
+					case <-ctx.Done():
+						ticker.Stop()
+						return
+					}
+				}
+			})
+			if err, ok := <-errChan; ok {
+				return err
+			}
+			// channel closed
+			return nil
+		}
+
+		notify := func(err error, time time.Duration) {
+			logger.Errorf("Provider connection error %w, retrying in %s", err, time)
+		}
+		err := backoff.RetryNotify(safe.OperationWithRecover(operation), backoff.WithContext(job.NewBackOff(backoff.NewExponentialBackOff()), ctxLog), notify)
+		if err != nil {
+			logger.Errorf("Cannot connect to http server %w", err)
+		}
+	})
+
+	return nil
+}
+
+// getDataFromEndpoint gets data from the configured provider endpoint, and returns the data.
+func (p *Provider) getDataFromEndpoint(ctx context.Context) ([]byte, error) {
+	b := []byte{}
+	req, err := http.NewRequest(http.MethodGet, p.Endpoint, bytes.NewBuffer(b))
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to create request: %w", err)
+	}
+
+	client := &http.Client{Timeout: time.Duration(p.PollTimeout)}
+	resp, err := client.Do(req)
+
+	if resp != nil {
+		defer resp.Body.Close()
+
+		var (
+			bodyData []byte
+			bodyErr  error
+		)
+
+		if bodyData, bodyErr = ioutil.ReadAll(resp.Body); bodyErr != nil {
+			return nil, fmt.Errorf("unable to read response body: %w", bodyErr)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("received non-ok response code: %d", resp.StatusCode)
+		}
+
+		log.FromContext(ctx).Debugf("Successfully received data from endpoint: %q", p.Endpoint)
+
+		return bodyData, nil
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to get data from endpoint: %w", err)
+	}
+
+	return nil, fmt.Errorf("received no data from endpoint")
```

Instead of checking if `resp != nil` above and if not to return an error at the end of the func, what do you think of failing fast and writing

```go
		if resp == nil {
			return nil, fmt.Errorf("received no data from endpoint")
		}

		defer resp.Body.Close()

		var (
			bodyData []byte
			bodyErr  error
		)

		if bodyData, bodyErr = ioutil.ReadAll(resp.Body); bodyErr != nil {
			return nil, fmt.Errorf("unable to read response body: %w", bodyErr)
		}

		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("received non-ok response code: %d", resp.StatusCode)
		}

		log.FromContext(ctx).Debugf("Successfully received data from endpoint: %q", p.Endpoint)

		return bodyData, nil
```

instead? This way the natural flow of the function is preserved, and the expected valid outcome is at the end of the function, instead of nested within an `if` statement